fhir-persistence-params/src/main/java/com/ibm/fhir/persistence/params/api/IParamValueCollector.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * (C) Copyright IBM Corp. 2022
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+ 
+package com.ibm.fhir.persistence.params.api;
+
+import com.ibm.fhir.persistence.exception.FHIRPersistenceException;
+import com.ibm.fhir.persistence.index.DateParameter;
+import com.ibm.fhir.persistence.index.LocationParameter;
+import com.ibm.fhir.persistence.index.NumberParameter;
+import com.ibm.fhir.persistence.index.ProfileParameter;
+import com.ibm.fhir.persistence.index.QuantityParameter;
+import com.ibm.fhir.persistence.index.ReferenceParameter;
+import com.ibm.fhir.persistence.index.SecurityParameter;
+import com.ibm.fhir.persistence.index.StringParameter;
+import com.ibm.fhir.persistence.index.TagParameter;
+import com.ibm.fhir.persistence.index.TokenParameter;
+
+/**
+ * Collects instances of search parameter values associated with a particular resource (identified by the logicalResourceId).
+ * This collector can be used to collect parameters from multiple resources being processed over an entire transaction.
+ * Collecting all the values in this way makes it more efficient when resolving reference (FK) ids, and means that
+ * we hold onto any locks for a shorter period of time.
+ */
+public interface IParamValueCollector {
+
+    /**
+     * Publish the values held in this collector using the given processor
+     * @param processor
+     */
+    void publish(IParamValueProcessor processor) throws FHIRPersistenceException;
+
+    /**
+     * Reset the state of the collector by clearing any values that have been previously collected
+     */
+    void reset();
+
+    /**
+     * To be called after all the values are resolved and the database transaction is committed.
+     * This call should promote any newly assigned database ids to the cache (it is then up
+     * to the cache implementation to decide what to do with those ids).
+     */
+    void publishValuesToCache();
+
+    /**
+     * Process the given LocationParameter p
+     * 
+     * @param tenantId
+     * @param requestShard
+     * @param resourceType
+     * @param logicalId
+     * @param logicalResourceId
+     * @param p
+     */
+    void collect(String tenantId, String requestShard, String resourceType, String logicalId, long logicalResourceId, LocationParameter p) throws FHIRPersistenceException;
```

Comment:
is "requestShard" here the same or different from our concept of "datastore id" (DSID)?

Reply:
This is part of the explicit sharding I introduced as a possible solution for Citus. It's currently a partial implementation because it looks like the DISTRIBUTED flavor will work for us. We may choose to remove it, but I don't want to do that now just in case we find out we need it as we scale out our performance testing of the DISTRIBUTED schema version.

Reply:
OK, thats what I thought.  But then I was wondering if we still support having multiple datastoreId because I didn't see it listed here.  Why include tenantId but not datastoreId?

Reply:
Assuming com.ibm.fhir.persistence.params.batch.ParameterValueCollector is the one implementation (I didn't see one in `fhir-remote-index`), the tenantId parameter doesn't actual seem to be used.  So basically I was just wondering if we could remove that or if we need it for some reason I'm not understanding.